state/consul/consul.go

```diff
@@ -0,0 +1,164 @@
+// ------------------------------------------------------------
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+// ------------------------------------------------------------
+
+package consul
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/dapr/components-contrib/state"
+	"github.com/hashicorp/consul/api"
+	"github.com/pkg/errors"
+)
+
+// StateStore is a state store implementation for HashiCorp Consul.
+type StateStore struct {
+	client        *api.Client
+	queryOpts     *api.QueryOptions
+	keyPrefixPath string
+}
+
+type consulConfig struct {
+	Datacenter    string `json:"datacenter"`
+	HTTPAddr      string `json:"httpAddr"`
+	ACLToken      string `json:"aclToken"`
+	Scheme        string `json:"scheme"`
+	KeyPrefixPath string `json:"keyPrefixPath"`
+}
+
+// NewConsulStateStore returns a new consul state store.
+func NewConsulStateStore() *StateStore {
+	return &StateStore{}
+}
+
+// Init does metadata and config parsing and initializes the
+// Consul client
+func (c *StateStore) Init(metadata state.Metadata) error {
+	consulConfig, err := metadataToConfig(metadata.Properties)
+	if err != nil {
+		return fmt.Errorf("couldn't convert metadata properties: %s", err)
+	}
+
+	var keyPrefixPath string
+	if consulConfig.KeyPrefixPath == "" {
+		keyPrefixPath = "dapr"
+	}
+
+	config := &api.Config{
+		Datacenter: consulConfig.Datacenter,
+		Address:    consulConfig.HTTPAddr,
+		Token:      consulConfig.ACLToken,
+		Scheme:     consulConfig.Scheme,
+	}
+
+	client, err := api.NewClient(config)
+	if err != nil {
+		return errors.Wrap(err, "initializing consul client")
+	}
+
+	var queryOpts *api.QueryOptions
+
+	c.client = client
+	c.queryOpts = queryOpts
+	c.keyPrefixPath = keyPrefixPath
+
+	return nil
+}
+
+func metadataToConfig(connInfo map[string]string) (*consulConfig, error) {
+	b, err := json.Marshal(connInfo)
+	if err != nil {
+		return &consulConfig{}, err
+	}
+
+	var config consulConfig
+	err = json.Unmarshal(b, &config)
+	if err != nil {
+		return &consulConfig{}, err
+	}
+
+	return &config, nil
+}
+
+// Get retrieves a Consul KV item
+func (c *StateStore) Get(req *state.GetRequest) (*state.GetResponse, error) {
+	var queryOpts *api.QueryOptions
+	if req.Options.Consistency == state.Strong {
+		queryOpts.RequireConsistent = true
+	}
+
+	resp, queryMeta, err := c.client.KV().Get(fmt.Sprintf("%s/%s", c.keyPrefixPath, req.Key), c.queryOpts)
+	if err != nil {
+		return nil, err
+	}
+
+	if resp == nil {
+		return &state.GetResponse{}, nil
+	}
+
+	return &state.GetResponse{
+		Data: resp.Value,
+		ETag: queryMeta.LastContentHash,
+	}, nil
+}
+
+// Set saves a Consul KV item
+func (c *StateStore) Set(req *state.SetRequest) error {
+	var reqValByte []byte
+	b, ok := req.Value.([]byte)
+	if ok {
+		reqValByte = b
+	} else {
+		reqValByte, _ = json.Marshal(req.Value)
+	}
+
+	keyWithPath := fmt.Sprintf("%s/%s", c.keyPrefixPath, req.Key)
+
+	_, err := c.client.KV().Put(&api.KVPair{
```

Comment:
Does the `Put` operation also take in a consistency model? (strong, for example)

Reply:
As far as I know, the `Put` operation currently doesn't support the consistency model.

https://www.consul.io/api/kv.html#create-update-key

Reply:
ok